src/section/section_chain.rs

```diff
@@ -0,0 +1,1114 @@
+    /// trusted. A chain is considered trusted only if at least on of the `trusted_keys` is on its
+    /// main branch.
+    ///
+    /// # Explanation
+    ///
+    /// Consider this chain that contains fork:
+    ///
+    /// ```ascii-art
+    /// A->B->C
+    ///    |
+    ///    +->D
+    /// ```
+    ///
+    /// Now if the only trusted key is `D`, then there is no way to prove the chain is trusted,
+    /// because this chain would be indistinguishable in terms of trust from any other chain with
+    /// the same general "shape", say:
+    ///
+    /// ```ascii-art
+    /// W->X->Y->Z
+    ///    |
+    ///    +->D
+    /// ```
+    ///
+    /// So an adversary is easily able to forge any such chain.
+    ///
+    /// When the trusted key is on the main branch, on the other hand:
+    ///
+    /// ```ascii-art
+    /// D->E->F
+    ///    |
+    ///    +->G
+    /// ```
+    ///
+    /// Then such chain is impossible to forge because the adversary would have to have access to
+    /// the secret key corresponding to `D` in order to validly sign `E`. Thus such chain can be
+    /// safely considered trusted.
+    pub fn check_trust<'a, I>(&self, trusted_keys: I) -> bool
+    where
+        I: IntoIterator<Item = &'a bls::PublicKey>,
+    {
+        let trusted_keys: HashSet<_> = trusted_keys.into_iter().collect();
+        let mut index = self.tree.len();
+
+        loop {
+            let (key, parent_index) = if index > 0 {
+                let block = &self.tree[index - 1];
+                (&block.key, Some(block.parent_index))
+            } else {
+                (&self.root, None)
+            };
+
+            if trusted_keys.contains(key) {
+                return true;
+            }
+
+            if let Some(next_index) = parent_index {
+                index = next_index;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    /// Compare the two keys by their position in the chain. The key that is higher (closer to the
+    /// last key) is considered `Greater`. If exactly one of the keys is not in the chain, the other
+    /// one is implicitly considered `Greater`. If none are in the chain, they are considered
+    /// `Equal`.
+    pub fn cmp_by_position(&self, lhs: &bls::PublicKey, rhs: &bls::PublicKey) -> Ordering {
+        match (self.index_of(lhs), self.index_of(rhs)) {
+            (Some(lhs), Some(rhs)) => lhs.cmp(&rhs),
+            (Some(_), None) => Ordering::Greater,
+            (None, Some(_)) => Ordering::Less,
+            (None, None) => Ordering::Equal,
+        }
+    }
+
+    /// Returns the number of blocks in the chain. This is always >= 1.
+    pub fn len(&self) -> usize {
+        1 + self.tree.len()
+    }
+
+    fn insert_block(&mut self, new_block: Block) -> usize {
+        // Find the index into `self.tree` to insert the new key at. All the keys above will be
+        // pushed upwards.
```

Comment:
Maybe the "All the keys above will be pushed upwards." bit can be removed as that's the default behaviour of `Vec::insert`?

Reply:
yeah the current sentence can be removed, as it's vectore::insert default behaviour.
However, it will be better to explain the `rule` used to find the insertion index ? 
i.e. something like the one I got above ?